### PR TITLE
refactor(ecam): wheel to react

### PR DESCRIPTION
**Pull request by davidwalschots**
## Summary of Changes
Rewrite the WHEEL page in React.

Includes some reuse of components created for the F/CTL page.

## Additional context

Discord username (if different from GitHub): SjotgunSjonnie#9623

## Testing instructions
The WHEEL ECAM page should work _approximately_ the same as it used to. Note that the spoiler indication now uses the same component as on the F/CTL page (this was out of sync before). Next to this, basic hydraulic availability is also used now. However, a future PR will further complete the HYD effect on this and other ECAM pages.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by davidwalschots** [comment]
Awaiting review on #4739. It contains components which I can reuse here.

File: src/instruments/src/SD/Common/HydraulicsProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+export const HydraulicsProvider = ({ children }) => {
+    const [greenPress] = useSimVar('L:A32NX_HYD_GREEN_PRESSURE', 'number', 1000);
+    const [yellowPress] = useSimVar('L:A32NX_HYD_YELLOW_PRESSURE', 'number', 1000);
+    const [bluePress] = useSimVar('L:A32NX_HYD_BLUE_PRESSURE', 'number', 1000);
+
+    const hydraulicContext: HydraulicsContext = {
+        G: { available: greenPress > 1450 },
+        Y: { available: yellowPress > 1450 },
+        B: { available: bluePress > 1450 },
+    };
```
Comment (crocket63) [inline src/instruments/src/SD/Common/HydraulicsProvider.tsx:28]: I think it should be a hysteresis 1450 / 1750 like ecam hyd ?
This should come as a boolean from hydraulic wasm to simplify the life of everyone....

Comment (lukecologne) [reply src/instruments/src/SD/Common/HydraulicsProvider.tsx:28]: The HYD ECAM SD page doesn't use any hysteresis for the system low pressure, it's just a simple comparison. If pressure<1450, the color is in amber, and else, it's green.

File: src/instruments/src/SD/Common/Spoilers.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    const [leftSpoilerState] = useSimVar('SPOILERS LEFT POSITION', 'percent over 100', 50);
+    const [rightSpoilerState] = useSimVar('SPOILERS RIGHT POSITION', 'percent over 100', 50);
+    const [speedBrakeHandlePosition] = useSimVar('SPOILERS HANDLE POSITION', 'percent over 100', 100);
+
+    const [spoilersArmed] = useSimVar('L:A32NX_SPOILERS_ARMED', 'boolean', 500);
+
+    const leftSpoilerUp = leftSpoilerState > 0.1;
+    const leftAileronUp = aileronLeftDeflectionState < -0.5;
+    const rightSpoilerUp = rightSpoilerState > 0.1;
+    const rightAileronUp = aileronRightDeflectionState > 0.5;
```
Comment (crocket63) [inline src/instruments/src/SD/Common/Spoilers.tsx:23]: I guess -0.5 and 0.5 are good values but I prefer to ask :D

Comment (davidwalschots) [reply src/instruments/src/SD/Common/Spoilers.tsx:23]: I have no idea. These values were in the old code and this PR (mostly) tries to refactor from old to React. I'm sure these values can or should be improved. As far as I could see right now these values show the joystick deflection, not the actual deflection of the various control surfaces. So yeah, that's likely wrong 😀.

File: src/instruments/src/SD/Pages/Wheel/Wheel.tsx [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+    return !hydraulics.G.available ? (
+        <SvgGroup x={x} y={y}>
+            <HydraulicIndicator x={0} y={0} type="Y" />
+
+            <text x={86} y={18} className="big-text color-green">ALTN BRK</text>
+            <AccumulatorOnly x={45} y={28} />
+        </SvgGroup>
+    ) : null;
+};
```
Comment (crocket63) [inline src/instruments/src/SD/Pages/Wheel/Wheel.tsx:134]: Actually pressure threshold for brake system switch is different from "hyd_available" thresholds.
// Feedback by Cpt. Chaos — 25/04/2021 #pilot-feedback
    const MIN_PRESSURE_BRAKE_ALTN_HYST_LO: f64 = 1305.;
    const MIN_PRESSURE_BRAKE_ALTN_HYST_HI: f64 = 2176.;

File: src/instruments/src/SD/Common/HydraulicsProvider.tsx (hunk 1 above)
Comment (davidwalschots) [reply src/instruments/src/SD/Common/HydraulicsProvider.tsx:28]: That could very well be. I've moved the values introduced in #4870 to a shared location. I would suggest creating a separate PR which modifies the WHEEL, F/CTL and HYD page to use calculated values coming from `systems.wasm` in a separate PR.

File: src/instruments/src/SD/Pages/Wheel/Wheel.tsx (hunk 3 above)
Comment (davidwalschots) [reply src/instruments/src/SD/Pages/Wheel/Wheel.tsx:134]: I would suggest creating a separate PR which modifies the WHEEL, F/CTL and HYD page to use calculated values coming from `systems.wasm` in a separate PR. This PR refactors from Asobo's JS code to our React implementation.

**Review by crocket63 — COMMENTED** [review]
Not an expert on this so basically just checked namings and couple of hydraulics remarks

File: src/instruments/src/SD/Pages/Wheel/Wheel.tsx [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        </g>
+    );
+};
+
+interface WheelArchProps extends ComponentPositionProps {
+    type: 'top' | 'bottom',
+    green?: boolean,
+    amber?: boolean
+}
+const WheelArch = ({ x, y, type, green, amber }: WheelArchProps) => {
```
Comment (Benjozork) [inline src/instruments/src/SD/Pages/Wheel/Wheel.tsx:274]:
```suggestion

const WheelArch = ({ x, y, type, green, amber }: WheelArchProps) => {
```

File: src/instruments/src/SD/Pages/Wheel/Wheel.tsx [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+                x={436}
+                y={318}
+                left={{ number: 3, temperature: temperatures[2], hottest: maxTemperatureIndex === 2 }}
+                right={{ number: 4, temperature: temperatures[3], hottest: maxTemperatureIndex === 3 }}
+            />
+        </EcamPage>
+    );
+};
+
+const NoseWheelSteering = ({ x, y }) => {
```
Comment (Benjozork) [inline src/instruments/src/SD/Pages/Wheel/Wheel.tsx:68]: Perhaps we could extract `x, y` to something like

`type TwoDimensionalPosition = { x: number, y: number }`

Comment (Benjozork) [reply src/instruments/src/SD/Pages/Wheel/Wheel.tsx:68]: Seems we are repeating it, untyped, a lot.

**Review by Benjozork — APPROVED** [review]

**Comment by frankkopp** [comment]
Quality Assurance Tester Report

Discord: Cdr_Maverick#6475
Object of testing: #4783
Tier of Testing: 1
Date: 28.5.2021

Testing Process: Full flight EDDS-EDDM / visual check of HYD page esp. during changes in hyd pressure, gear position and break temp

Negatives: None

Testing Results: Passed

Conclusions: LGTM

**Comment by TheSquirtingPantin** [comment]
Quality Assurance Tester

Discord: Remi#8871
Object of testing: #4847
Tier of Testing: 1
Date: 28/05/2021

Testing Process:
1) Full Flight (KDEN-KIAD)
2) Checked WHEEL page on the ground and in-flight.Gear Extension/Retraction sequence checked, brake temp logic, available amber messages checked (LDG CTL, ANTI SKID...)

Negatives:  (Potentially in PR spectrum) LDG CTL message seems to be incorrectly displayed. The message is displayed each time the landing gear lever is selected up and/or down rather than when configuration is actually incorrect. Pretty sure this message is not shown when gear is in the transition phase during normal operations.

Testing Results: Passed. Good Job, page is similar to previous version (non REACT). Check remark for LDG CTL message logic (potentially within PR spectrum) 

Conclusions: LGTM. 
Media: N/A

**Comment by davidwalschots** [comment]
> Negatives: (Potentially in PR spectrum) LDG CTL message seems to be incorrectly displayed. The message is displayed each time the landing gear lever is selected up and/or down rather than when configuration is actually incorrect. Pretty sure this message is not shown when gear is in the transition phase during normal operations.

Yeah. I can imagine this is incorrect. However, as you noted yourself, indeed this is the behaviour that was already there in the non-React ECAM. We'll fix it some other time in a separate PR I'm sure! 😃

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/A320_Neo_EICAS.html [hunk 6]
```diff
@@ -52,7 +52,8 @@
                 <a320-neo-lower-ecam-bleed style="display: none"></a320-neo-lower-ecam-bleed>
                 <a320-neo-lower-ecam-press style="display: none"></a320-neo-lower-ecam-press>
                 <a32nx-door-page-element style="display: none"></a32nx-door-page-element>
-                <a320-neo-lower-ecam-wheel style="display: none"></a320-neo-lower-ecam-wheel>
+                <a32nx-fctl-page-element style="display: none"></a32nx-fctl-page-element>
```
Comment (lukecologne) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/A320_Neo_EICAS.html:55]: `a32nx-fctl-page-element` is duplicated

Comment (davidwalschots) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/A320_Neo_EICAS.html]: w00ps.

**Comment by DavidEGTE** [comment]
Quality Assurance Trainee Report

Discord: David#2177
Object of testing: #4783 
Tier of Testing: 1
Date: 28/05/2021

Testing Process:  Full flight LIRN to LIRF + visual check of HYD page
Negatives: N/A
Testing Results: Passed

Conclusions: LGTM

**Pull request merged**